lms/services/exceptions.py

```diff
@@ -220,6 +220,36 @@ def __init__(self, file_id):
         super().__init__(self.details)
 
 
+class CanvasGroupSetNotFound(ServiceError):
+    """A Canvas GroupSet not found on Canvas API."""
+
+    error_code = "canvas_group_set_not_found"
+
+    def __init__(self, group_set):
+        self.details = {"group_set": group_set}
+        super().__init__(self.details)
+
+
+class CanvasGroupSetEmpty(ServiceError):
+    """Canvas GroupSet doesn't contain any groups."""
+
+    error_code = "canvas_group_set_empty"
+
+    def __init__(self, group_set):
+        self.details = {"group_set": group_set}
+        super().__init__(self.details)
+
+
+class CanvasStudentNotInGroup(ServiceError):
+    """Student doesn't belong to any of the groups in a group set assignment."""
+
+    error_code = "canvas_student_not_in_group"
+
+    def __init__(self, group_set):
+        self.details = {"group_set": group_set}
+        super().__init__(self.details)
```

Comment:
:+1: I think I like the use of lots of small and separate exception classes, rather than all the same class with different `error_code`'s. It makes it easier to write `except`'s that only catch one type of exception or that treat different types of exceptions differently, either in code or in tests. The disadvantage of course is a lot of exception classes.

You might want these three to all share a common based class (`CanvasGroupError` or something) so that an `except` can more easily catch all three at once if it wants to do that. That might also help to remove some duplicate code that these three classes share.

(Generally I think deep inheritance hierarchies are super bad but exception classes, which should be very simple, are one of the places where it's okay.)

Good docstrings

Reply:
These exceptions probably don't belong in `services/` though, since they're raised by view code not services code